src/liballoc/vec.rs

```diff
@@ -2691,7 +2691,10 @@ impl<'a, T, F> Iterator for DrainFilter<'a, T, F>
                     self.del += 1;
                     return Some(ptr::read(&v[i]));
                 } else if self.del > 0 {
-                    v.swap(i - self.del, i);
+                    let del = self.del;
+                    let src: *const T = &v[i];
+                    let dst: *mut T = &mut v[i - del];
+                    ptr::copy_nonoverlapping(src, dst, 1);
```

Comment:
I feel like this requires a comment to explain why this is safe (for example in case of unwinding). The crutch is quite literally not local, which is that the Vec had its length set to 0.

Do the bounds checks optimize out here? Is there anything to gain from writing this entirely in raw pointers instead of using indices? We've seen before that the compiler doesn't optimize out indices and lengths into slices (or into slice iterators, not relevant here) entirely fluently.

Reply:
I'll add that comment.  The first implementation did use raw pointers but they did not perform favorably in benchmarks.  I believe this was because `pointer::offset` required me to convert the usize to an isize.  In order to prevent this I could just store isize values and instead make all usages of this use raw pointers.  I'll try that and see how it goes in benchmarking.